expertsystem/topology.py

```diff
@@ -41,22 +41,23 @@ def __eq__(self, other: object) -> bool:
         raise NotImplementedError
 
 
-class StateTransitionGraph:
-    """Graph class that contains edges and nodes.
+class Topology:
+    """Directed graph for a state transition.
 
-    Similar to feynman graphs. The graphs are directed, meaning the edges are
-    ingoing and outgoing to specific nodes (since feynman graphs also have a
-    time axis) This class can contain the full information of a state
-    transition from a initial state to a final state. This information can be
-    attached to the nodes and edges via properties.
+    Forms the underlying topology of `StateTransitionGraph`.
     """
 
-    def __init__(self) -> None:
+    def __init__(
+        self,
+        nodes: Optional[Set[int]] = None,
+        edges: Optional[Dict[int, Edge]] = None,
+    ) -> None:
         self.__nodes: Set[int] = set()
         self.edges: Dict[int, Edge] = {}
-        self.node_props: Dict[int, dict] = {}
-        self.edge_props: Dict[int, dict] = {}
-        self.graph_element_properties_comparator: Optional[Callable] = None
+        if nodes is not None:
+            self.__nodes = set(nodes)
+        if edges is not None:
+            self.edges = edges
```

Comment:
This unsafe as the edge could point to non existing nodes etc

Reply:
#250